### PR TITLE
fix: split over-long Slack replies before msg_too_long; diagnose shared-app event stealing (1.0.2)

### PR DESCRIPTION
**Pull request by anaclumos**
Two live serve incidents from 2026-07-20, both fixed here; bumps 1.0.2.

## Bug 1: replies died with `msg_too_long` (three lost turns)

Slack rejects an over-long message with `msg_too_long`, and nothing in chat 4.34.0 or `@chat-adapter/slack` bounds, truncates, or splits reply text (verified in-source): a natively streamed message accumulates server-side toward the documented 12,000-char `markdown_text` envelope (docs.slack.dev documents that limit on `chat.postMessage`/`update` and all three streaming methods), and once anything rendered natively a failed append REJECTS the whole `thread.post` - the reply vanished and the thread only got the "could not be posted" diagnostic.

Fix: `relayThread` splits reply text across Slack messages at `SEGMENT_TEXT_MAX` (10,000 chars; the 2,000-char margin absorbs the renderer's markdown normalization and mention-linkification expansion). A break prefers the last newline in the back half of the remaining room (a front-half newline followed by one giant unbroken run makes no progress and would loop against the fence-reopen prefix), and a break forced inside a code fence closes it and reopens it in the next message so both halves render as code. The no-stream `result` fallback goes through the same pump.

Known accepted tradeoff (documented in-file): the adapter-internal plain-text fallback edits via `chat.update` `text` (hard 4,000-char cap) but only engages when the workspace refused native streaming outright; splitting every normal reply 3x tighter to cover that never-hit path is worse than the residual risk.

## Bug 2: `serve.unlinked_channel` + socket-mode retry spam

Root cause is topology, not code: two hosts ran serve daemons on ONE shared Slack app, and Slack Socket Mode delivers each envelope to exactly one of an app's open connections ("It's best not to assume any particular pattern" - docs.slack.dev/apis/events-api/using-socket-mode). The daemons steal each other's events: mentions survive only because `app_mention` + `message.channels` are two envelopes (the Chat SDK dedupes them per-process by ts), but a non-mention thread follow-up rides ONE envelope and silently dies when the wrong daemon receives it (unsubscribed there = debug-level drop inside chat core). The retry storm in the log was Slack redelivering unacked envelopes from a restart gap, amplified across both daemons.

No code can route Slack's load-balancing: the hello frame's `num_connections` is discarded by `@slack/socket-mode` 2.0.7 before any event fires, and the adapter's `socketClient` is typed private. So the fix is diagnosis + docs:

- `serve.unlinked_channel` logs ONCE per channel per daemon run, with the shared-app explanation in the log line (was: one line per redelivered envelope).
- The adapter gets an explicit warn-level `ConsoleLogger` (its info-level default printed `Processing socket mode retry` per envelope; warn matches the Chat core logger, and real degradations like streaming fallback are warn-level and survive).
- `docs/limitations.mdx` + `AGENTS.md` record the one-Slack-app-per-daemon constraint.

Operational follow-up for the owner (not in this PR): the second host needs its own Slack app - paste the `serve setup` manifest as a new app with a distinct bot display name and re-run `serve setup` there with the new tokens.

## Tests

- relayThread splitting: ordered posts for a 25k reply, delta accumulation across the cap, code-fence close/reopen continuity, huge no-stream result splitting.
- serve runtime: unlinked-channel logs once per channel with the diagnosis.
- Full suite: 412 pass / 0 fail; `bun run typecheck` clean.

---

> [!NOTE]
> **Medium Risk**
> Changes the Slack relay posting path for all long replies and salvage recovery; behavior is heavily tested but affects production serve threads.
> 
> **Overview**
> Fixes two **2026-07-20** serve incidents and bumps **1.1.2**.
> 
> **Long replies no longer die with `msg_too_long`.** `relayThread` routes streamed text and no-stream `result` bodies through **`pushText`**, splitting at **`SEGMENT_TEXT_MAX` (10,000)** into ordered Slack messages before Slack’s ~12k `markdown_text` cap. Breaks prefer a late newline in the remaining room, avoid cutting through backtick runs, and close/reopen fenced code when a cut lands inside a fence (parity tracked on **accumulated** segment text, including deltas that split ``` across chunks). Post-failure **salvage** now propagates open-fence state via **`reopenFence`** so recovered text still renders as code (including card-only and chained salvages).
> 
> **Shared Slack app topology is documented and easier to spot.** `serve.unlinked_channel` logs **once per channel per daemon run** with a note that Socket Mode delivers each envelope to only one connection (thread follow-ups vanish when the wrong host wins). The Slack adapter uses a **warn-level `ConsoleLogger`** to cut Socket Mode retry noise. **`docs/limitations.mdx`** and design docs state **one Slack app per daemon**.
> 
> Tests cover splitting, fence/salvage edge cases, and the unlinked-channel log behavior.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 52cc3431fcd061079b17131c717fdc8fd26f0209. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

---
## Summary by cubic
Splits over-long Slack replies before Slack’s cap and adds once-per-channel logging to diagnose shared-app event stealing, releasing 1.1.2. Splitting is fence‑safe across streamed deltas, no‑stream results, and salvage, including card‑only and chained salvages.

- **Bug Fixes**
  - Split reply text at 10,000 chars via a shared `pushText` path for streamed deltas and no‑stream results; prefers newline breaks; computes code‑fence parity over accumulated text; never cuts through backtick runs; closes/reopens fences when a cut lands inside code.
  - Salvage reopens a fence that the delivered prefix left open and holds a pending reopen until the first later text enters the follow‑on message; propagates a still‑armed reopen across chained salvages so later text renders as code.
  - Diagnose shared Slack‑app setups: `serve.unlinked_channel` logs once per channel per daemon run with an explanation; `@chat-adapter/slack` uses a warn‑level `ConsoleLogger` from `chat` to reduce Socket Mode retry noise; docs note “one Slack app per daemon”.
  - Tests cover ordered splitting (large texts and accumulating deltas), fence continuity across delta boundaries, backtick‑run safety, salvage fence reopen (card‑only and chained), no‑stream result splitting, and the once‑per‑channel log.

<sup>Written for commit 52cc3431fcd061079b17131c717fdc8fd26f0209. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/anaclumos/tokenmaxxing/pull/42?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

File: src/lib/slackbridge.ts [hunk 1]
```diff
@@ -437,13 +454,46 @@ export async function relayThread(input: {
     if (!(chunk instanceof Object)) {
       postedText = true;
       segmentMeta!.text = true;
+      segmentMeta!.chars += chunk.length;
+      // fence parity by occurrence count: an odd number of ``` markers in this
+      // chunk flips whether the segment currently sits inside a code fence.
+      if ((chunk.split("```").length - 1) % 2 === 1) segmentMeta!.fenceOpen = !segmentMeta!.fenceOpen;
```
[the remark this answers is not in the document]
Comment (anaclumos) [reply src/lib/slackbridge.ts]: Agreed and fixed in 3f97de6: fence parity is now computed over the segment's ACCUMULATED text (segmentMeta.acc glues deltas back together), so a ``` delimiter split across two SDK deltas tracks correctly. Regression test added ("a fence delimiter split across SDK deltas still tracks as one fence").

File: src/lib/slackbridge.ts [hunk 2]
```diff
@@ -437,13 +454,46 @@ export async function relayThread(input: {
...
+      }
+      if (rest.length <= room) {
+        await push(rest);
+        return;
+      }
+      // prefer a newline cut only when it lands in the back half of the room:
+      // an early newline followed by one giant unbroken run would otherwise
+      // make no progress and (with the fence-reopen prefix) loop forever.
+      const nl = rest.lastIndexOf("\n", room - 1);
+      const head = rest.slice(0, nl >= Math.floor(room / 2) ? nl + 1 : room);
```
[the remark this answers is not in the document]
Comment (anaclumos) [reply src/lib/slackbridge.ts]: Agreed and fixed in 3f97de6: the cut now detects a backtick run straddling the boundary and walks left past the whole run, so a partial delimiter is never stranded on either side. The one exception is pathological all-backtick input reaching position 0, where the original cut is kept (progress beats rendering there). Regression test added ("a hard cut never slices through a backtick run").

File: docs/content/docs/limitations.mdx [hunk 3]
```diff
@@ -8,4 +8,5 @@ description: The tradeoffs that come with pooled hot-swapping
 - **Adoption lag.** On macOS the first turn within ~30 seconds of a swap can still meter the old account; the bars' headroom absorbs it.
 - **Shared blast radius.** All default-profile sessions share one live credential, so a swap moves them all together (each adopts in place).
 - **Plaintext credentials on Linux.** Claude Code itself stores Linux credentials as a 0600 plaintext file; tokenmaxxing's parked copies follow the same model.
+- **One Slack app per serve daemon.** Slack Socket Mode delivers each event to only one of an app's open connections, so two `xx serve` daemons sharing one Slack app steal events from each other: mentions usually survive (Slack emits two events per mention), but plain thread replies ride a single event and silently vanish whenever the other host receives it. Give every host its own Slack app with its own tokens; a daemon that sees traffic for a channel it does not serve logs `serve.unlinked_channel` with this diagnosis.
```
[the remark this answers is not in the document]
Comment (anaclumos) [reply docs/content/docs/limitations.mdx:11]: Partially agreed. DESIGN.md now carries both the reply-splitting mechanics and the one-Slack-app-per-daemon constraint (3f97de6, section 5b) - operating docs move with operational-semantics changes. The 29 localized limitations pages are deliberately not hand-edited in this PR: the English page is the source of truth and locale files are regenerated in dedicated bulk docs passes (see #26, which produced all 30 locales in one sweep); partial hand translations drift against the next sweep.

File: src/lib/slackbridge.ts [hunk 4]
```diff
@@ -561,8 +583,15 @@ export async function relayThread(input: {
...
-      // result-only answer (adversarial-review catch on PR #45).
-      if (!(chunk instanceof Object)) meta.text = true;
+      // result-only answer (adversarial-review catch on PR #45). acc still
+      // accumulates: it mirrors the segment's FULL text on every entry path,
+      // so pushText's room accounting and fence parity see salvaged text too
+      // (a salvage segment that continued via pushText could otherwise grow
+      // past the msg_too_long cap).
+      if (!(chunk instanceof Object)) {
+        meta.text = true;
+        meta.acc += chunk;
```
[the remark this answers is not in the document]
Comment (anaclumos) [reply src/lib/slackbridge.ts:622]: Agreed, fixed in b745650: the rejection handler now computes fence parity over the DELIVERED prefix and splices a ```-reopen before the first salvaged text piece, mirroring pushText's reopen at a split boundary. The reopen rides pushInto so the salvage segment's acc parity matches what Slack displays, and your combined-coverage ask is pinned by a renderer-faithful regression: eager open-fence commits (verified against the real StreamingMarkdownRenderer: fenced content commits aggressively, which is why Slack streams code live), a two-delta death mid-fence, and the salvage message beginning inside a fence with every piece delivered exactly once. The dead message's own unclosed fence is accepted cosmetic damage, documented in-file.

File: src/lib/slackbridge.ts [hunk 5]
```diff
@@ -541,6 +563,18 @@ export async function relayThread(input: {
...
+        // A fence OPENED in the delivered prefix leaves the remainder's code
+        // fenceless in the fresh salvage message (pullfrog catch, PR #42):
+        // reopen it before the first salvaged text piece, exactly like
+        // pushText's reopen at a split boundary. The reopen rides pushInto,
+        // so the salvage segment's acc parity matches what Slack displays.
+        // The DEAD message's own unclosed fence is accepted cosmetic damage:
+        // nothing can append to a rejected stream.
+        const deliveredFenceOpen = (fullRaw.slice(0, deliveredLen).split("```").length - 1) % 2 === 1;
+        if (deliveredFenceOpen) {
+          const firstText = lost.findIndex((c) => !(c instanceof Object));
```
[the remark this answers is not in the document]
Comment (anaclumos) [reply src/lib/slackbridge.ts]: Agreed, fixed in the just-pushed commit: the eager splice became a PENDING reopen armed on the salvage segment's meta and materialized right before the first text entering the segment from either entry point (salvaged remainder via pushInto, or a later streamed push joining the open segment). A card-only salvage no longer skips the reopen, and a text-less salvage never dangles an empty open fence. Your card-then-later-text ordering is pinned by a new regression.

File: src/lib/slackbridge.ts [hunk 6]
```diff
@@ -541,6 +570,13 @@ export async function relayThread(input: {
             rest = rest.slice(nl + 1);
           }
         }
+        // A fence OPENED in the delivered prefix leaves later code fenceless
+        // in the fresh salvage message (pullfrog catches, PR #42): arm the
+        // segment's pending reopen, materialized right before the FIRST text
+        // that enters it - whether a salvaged remainder piece here or a later
+        // streamed push joining the segment (a card-only salvage must not
+        // skip the reopen, and a text-less segment must not dangle one).
+        const deliveredFenceOpen = (fullRaw.slice(0, deliveredLen).split("```").length - 1) % 2 === 1;
```
[the remark this answers is not in the document]
Comment (anaclumos) [reply src/lib/slackbridge.ts]: Agreed, fixed in 52cc343 with exactly your suggested XOR (comment wording adopted nearly verbatim): the dying segment's still-armed reopenFence folds into the parity, materialization makes it a no-op, and a normal segment's flag is false. The chained double-death regression pins the reopen landing before the later text.

[the remark this answers is not in the document]
Comment (anaclumos) [reply src/lib/slackbridge.ts]: Agreed, fixed in 52cc343 (same fix as vercel's suggestion on this line): meta.reopenFence is preserved by XOR-folding it into the next salvage's deliveredFenceOpen. Pinned by the chained card-only double-death regression.
